052-k8s-server-side-apply.md

```diff
@@ -0,0 +1,77 @@
+
+# Kubernetes Server Side Apply
+
+Make the Operators use [Kubernetes Server Side Apply](https://kubernetes.io/docs/reference/using-api/server-side-apply). This will enable third parties (e.g. Kyverno, Kopf) to have their own labels/annotations on Strimzi owned Kubernetes Resources without Strimzi attempting to remove them.
+
+NOTE: Strimzi will still have ownership over the fields it wants.
+
+## Current situation
+
+As per [Issue 6938](https://github.com/strimzi/strimzi-kafka-operator/issues/6938) third party Kubernetes operators are having a negative interaction with the existing way Strimzi does Kubernetes applies.
+An example is [Kyverno](https://kyverno.io/) has a policy in a K8s cluster to add a label to all resources with a MutatingWebhookConfiguration, Strimzi detects this label as a resource diff and removes it as it doesn't own it / not included in [hard-coded list of ignored labels](https://github.com/strimzi/strimzi-kafka-operator/blob/c3522cf4b17004004a676854d37ba01bb9a44800/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetDiff.java#LL28C34-L28C49). This triggers the MutatingWebhookConfiguration which re-applies the label and the loop continues.
+
+## Motivation
+
+Once this change is made (which is relatively simple to implement), it will allow third party tools within the cluster to interact with Strimzi owned resources without looping and Strimzi will be operating as if it never knew about them.
+
+## Proposal
+
+This proposal makes Kubernetes keep track of each field's "owner" and any changes submitted to those fields will be compared to the remote state before being applied. 
+
+This would let the Kubernetes apiserver handle the three-way diff, updates etc, and allow other controllers, webhooks etc to touch fields, labels, annotations etc the operator doesn't care about.
+
+Conflicts can happen with Server Side Apply as described [here](https://github.com/fabric8io/kubernetes-client/blob/v6.5.1/doc/CHEATSHEET.md#server-side-apply), it is open to choose the method that Strimzi would use for this.
+
+[Kubernetes suggest](https://kubernetes.io/docs/reference/using-api/server-side-apply/#conflicts:~:text=It%20is%20strongly%20recommended%20for%20controllers%20to%20always%20%22force%22%20conflicts%2C%20since%20they%20might%20not%20be%20able%20to%20resolve%20or%20act%20on%20these%20conflicts.) using the above _force_ method to make sure Strimzi have control over the parts they want. As described in the Kubernetes docs _"This forces the operation to succeed, changes the value of the field, and removes the field from all other managers' entries in managedFields"_, this obviously means if another party is _forcing_ the same field that they will create a loop of ownership but this would be rare as operators should only care about their own fields.
```

Comment:
So, how will the operator detect and force the change? You are quoting from the Kube docs, but that is just a theoretical document. How would this be handled in practice? Will we force everything? Will we detect the issue and force? Does the Kubernetes client handle this?

Reply:
When the operator does a server-side apply, if there are no conflicts, whether you set force or not doesn't matter.

If you set force, then your changes will be applied (and managedFields updated to you), whether or not there was a conflict. If you don't set force, then it will error out if there's a conflict.

All of the above is handled automatically by the server.

How does the operator detect this? It doesn't really care - ResourceDiff still controls whether it tries to do an apply or not, so if it sees something out of sync it cares about it'll do the apply and overwrite any conflicting changes

Reply:
No - The proposal says that we should set force to be true all the time. ResourceDiff doesn't handle conflicts, or control whether force is used or not

ResourceDiff is still used to determine whether or not to do an apply, purely based on the fields it has in desired (but not looking at any other fields of the object)

tbh I'm not sure why we're needing to discuss the actual code changes in the proposal instead of in an actual PR to change the code.

Reply:
> AFAIU, it is proposed that: ...

Yes, that's my current understanding

> Or to put it another way, the cost of a no-op reconcile is still a single GET of the current resource.

Yes

> One very minor change is that (I assume) the apiserver needs a few more cycles to handle a server-side apply, as it has to compute the diff and detect conflicts itself. Right now the kube client sends a JSON patch, rather than the full resource state when using the JSON patch strategy. But the cost is offset because the kube client doesn't have to do it's own GET to compute that JSON patch.

I feel this is going to be an insignificant cost in the context of an operator, and the number of updates/changes that are happening. I think the apiserver still does have to do the work of diffing and detecting conflicts - it needs to update `managedFields` still, which needs to remain consistent with the object state, even if actors are proposing direct changes. (If you're [making non-server-side-apply updates](https://kubernetes.io/docs/reference/using-api/server-side-apply/#apply-and-update), the apiserver will still be updating managedFields, and guessing a manager name from your HTTP user-agent)

> A consequence of 3 is that the Strimzi operator would never know (and by extension the person changing a Strimzi CR may never know) that some other operator/hook's preferred value for a field is being overwritten. Indeed it no one it looking at the output of that Other Thing such a conflict could ping-pong indefinitely. It's easy to make assertions like "this would be rare as operators should only care about their own fields", but in practice those come back to us as bug reports. 

Yeah, and this is actually the current behaviour - The current operator behaviour doesn't know that some other operator/hook/standard k8s machinery is being overwritten, it's overwriting all fields all the time.

We have a few examples of such bug reports:
- https://github.com/orgs/strimzi/discussions/8667 (pvc resizing)
- https://github.com/strimzi/strimzi-kafka-operator/issues/4035 (cattle annotations)
- https://github.com/strimzi/strimzi-kafka-operator/issues/6938 (keiko iam controller, metallb, kopf)

My hypothesis is that there will be fewer bug reports in a world where strimzi uses server-side-apply, than in the current world where it does not

> So I wonder if we make a first request with force=false, log a warning (or similar) if there's a conflict and then retry with force=true. That at least makes the game of ping pong more apparent.

Mhm, that's reasonable

Reply:
Yes, you can use APPLY to create (create if not exists or update)
It depends on what triggers the reconcile, I guess?
- You could do an APPLY on being told about an upstream resource change (by watch etc)
  - This is also on startup, with the initial watch? (Haven't checked this is how the code works, just assuming)
- If we're watching the downstream resources too, can also fire off an APPLY when they change

If something downstream is making changes regularly (HPA? .status updates on all sorts of objects?) you may end up with lots of applies happening. But if they're going to be no-ops (because all of the fields set in the `desired` are already correct), then this may not matter?

Reply:
[This blog by Kubernetes](https://kubernetes.io/blog/2022/10/20/advanced-server-side-apply/#reconstructive-controllers) goes into this a bit more